core/arch/riscv/kernel/thread_rv.S

```diff
@@ -0,0 +1,279 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2022 NXP
+ */
+
+#include <asm.S>
+#include <generated/asm-defines.h>
+#include <keep.h>
+#include <kernel/thread.h>
+#include <kernel/thread_private.h>
+#include <mm/core_mmu.h>
+#include <riscv.h>
+#include <riscv_macros.S>
+
+.macro save_regs, mode
+	addi	sp, sp, -RISCV_TRAP_REGS_SIZE
+.if \mode == TRAP_MODE_USER
+
+	store_xregs	sp, RISCV_TRAP_REG_TP, 4, 4
+	load_xregs	sp, (RISCV_TRAP_REGS_SIZE - RISCV_XLEN_BYTES), 4, 4
+
+	store_xregs	sp, RISCV_TRAP_REG_GP, 3, 3
+	csrrw	gp, CSR_XSCRATCH, zero
+	store_xregs	sp, RISCV_TRAP_REG_SP, 3, 3
+.endif
+
+	store_xregs	sp, RISCV_TRAP_REG_T3, 28, 31
+	store_xregs	sp, RISCV_TRAP_REG_A6, 16, 17
+	store_xregs	sp, RISCV_TRAP_REG_T0, 5, 7
+	store_xregs	sp, RISCV_TRAP_REG_A0, 10, 15
+
+	store_xregs	sp, RISCV_TRAP_REG_RA, 1, 1
+
+	csrr	x5, CSR_XSTATUS
+	store_xregs	sp, RISCV_TRAP_REG_STATUS, 5, 5
+
+	csrr	a0, CSR_XCAUSE
+	csrr	a1, CSR_XEPC
+
+	store_xregs	sp, RISCV_TRAP_REG_EPC, 11, 11
+
+	mv	a2, sp
+
+	/* a0 = cause
+	 * a1 = epc
+	 * a2 = sp
+	 * a3 = user
+	 * thread_trap_handler(cause, epc, sp, user)
+	 */
+.endm
+
+.macro restore_regs, mode
+
+	load_xregs	sp, RISCV_TRAP_REG_EPC, 5, 5
+
+	csrw	CSR_XEPC, t0
+
+	load_xregs	sp, RISCV_TRAP_REG_STATUS, 5, 5
+	csrw	CSR_XSTATUS, t0
+
+	load_xregs	sp, RISCV_TRAP_REG_RA, 1, 1
+
+	load_xregs	sp, RISCV_TRAP_REG_A0, 10, 15
+	load_xregs	sp, RISCV_TRAP_REG_T0, 5, 7
+
+	load_xregs	sp, RISCV_TRAP_REG_A6, 16, 17
+	load_xregs	sp, RISCV_TRAP_REG_T3, 28, 31
+
+.if \mode == TRAP_MODE_USER
+	addi	gp, sp, RISCV_TRAP_REGS_SIZE
+	store_xregs	gp, REGOFF(-1), 4, 4
+	csrw	CSR_XSCRATCH, gp
+
+	load_xregs	sp, RISCV_TRAP_REG_TP, 4, 4
+	load_xregs	sp, RISCV_TRAP_REG_GP, 3, 3
+	load_xregs	sp, RISCV_TRAP_REG_SP, 2, 2
+
+.else
+	addi	sp, sp, RISCV_TRAP_REGS_SIZE
+.endif
+.endm
+
+/* size_t __get_core_pos(void); */
+FUNC __get_core_pos , : , .identity_map
+	lw	a0, THREAD_CORE_LOCAL_HART_ID(tp)
+	ret
+END_FUNC __get_core_pos
+
+FUNC thread_trap_vect , :
+	csrrw   sp, CSR_XSCRATCH, sp
```

Comment:
hello, I saw you using sscratch register to judge from kernel code or user application, but sscratch register have no initial value？

Reply:
If it has an initial value how to distinguish between traps coming from user and those coming from kernel ? Here we are not reading xscratch, we are swapping sp and xscratch, we don't restore xscratch if trap comes from user. Otherwise xscratch is used initially by the boot code to store the hart ID.

Reply:
hi maroueneboubakri, 
thanks for your quickly repsonse. I know you are swapping sp and xscratch. when xscratch is not zero, "bnez sp, 0f" will jump to trap_from_user, then using sp(xscratch) as stack area to save register.If xscratch is not point to a vaild stack area, save_regs macro code will work wrong?

Reply:
Hi @maroueneboubakri
I think the concern from @matthewgui is reasonable. xscratch need a initial value.
1. Initialize xscratch as zero in ``thread_init_per_cpu()``.
2. Assign non-zero value to xscratch in ``__thread_enter_user_mode()`` to indicate the thread will run as user mode.
3. Zero the xscratch in ``thread_unwind_user_mode()`` to re-enter kernel mode.

Reply:
Hi @matthewgui, @gagachang, Sure, I'll look closer into this, I'll rely more in `xstatus.SPP`.

Reply:
That would be better!! xstatus.SPP is more precise.

Reply:
BTW, gp can be used in linker relaxation for optimization.
It should be fixed value after assigning  __global_pointer$ to gp.

Reply:
@matthewgui, @gagachang, comments addressed, check:

- https://github.com/OP-TEE/optee_os/blob/9a3167969650dfbe0ce131d95301896e4c00062a/core/arch/riscv/kernel/thread.c#L563-L568
- https://github.com/OP-TEE/optee_os/blob/9a3167969650dfbe0ce131d95301896e4c00062a/core/arch/riscv/kernel/thread_rv.S#L26-L30
- https://github.com/OP-TEE/optee_os/blob/9a3167969650dfbe0ce131d95301896e4c00062a/core/arch/riscv/kernel/thread_rv.S#L158
- https://github.com/OP-TEE/optee_os/blob/9a3167969650dfbe0ce131d95301896e4c00062a/core/arch/riscv/kernel/thread_rv.S#L226

Thanks

Reply:
Dear @maroueneboubakri ,

Linux kernel also uses sscratch CSR to distinguish whether trap is from kernel or user.
https://elixir.bootlin.com/linux/latest/source/arch/riscv/kernel/entry.S#L27

If we use sscrach in OP-TEE too, we might break the kernel side, am I correct?

Reply:
No, when you switch-context to OP-TEE from Linux and vice versa (in case you are relying open a SM) the SM will store the whole context(including sscratch) and restore it later when back. In case you are using domains, sscratch is per-hart register therefore each OS will have its `own` sscratch.

Reply:
Yes, how about the linux-side interrupts.
If some interrupts are delegated to S-mode linux, the SM will be skipped and the interrupts are directly injected into S-mode.
Is it possible that a linux-side interrupt preempts OP-TEE side U-mode thread, therefore linux gets wrong sscratch value?

Reply:
Dear @maroueneboubakri ， 
as @gagachang  mentioned interrupts，when cpu run on optee world，if local timer happens，what should optee os do？For ARM，maybe system will forward to normal world to deal with localtimer interrupt。I see your code only clear TIE interrupt. Do you have  a secure and non-secure interrupt handling plans? 
thanks!

Reply:
@matthewgui, the peripheral isolation needs IOPMP support. But it has not been ratified yet.

For interrupt delegation, it can be forbidden or ignored in this stage.
SM can handle all asserted interrupts and redirect them to S-mode Linux or OP-TEE respectively with context switch. Therefore the context of them could both be correct.
But SM also needs some mechanisms to distinguish the asserted interrupt belongs to which OS.

Reply:
@gagachang , yes, i agree with you, SM needs some mechanisms to distinguish the asserted interrupt security state。 Compared with arm，riscv still need some security hardware mechanisms。
thanks！

Reply:
@matthewgui @gagachang Please note that we are not considering using the SBI/ECALL based conduit method like SMC in ARM. No SM, Instead we consider messaging between domains. As mentioned [here](https://github.com/OP-TEE/optee_os/issues/5676). We will add a helper function to control the execution flow in case of using SBI/ECALL based transport. It seems that @liushiwei007 has something to propose in this direction. See [here](https://github.com/riscv-software-src/opensbi/issues/282). In the first port we made we rely on Keystone SM to handle context switching between Linux and OP-TEE enclaves. But later we decided to withdraw that due many constraints like the lack of hardware-enforced mechanisms.

Reply:
@maroueneboubakri  Thanks for explanation.
Yes you are right, AMP might be better solution since current RISC-V lacks secure extensions such as secure-aware interrupt controller. We are glad to see many solutions. Please feel free to adopt your solution and just keep some software abstraction so that communities can choose their own transport method between RPMSG and SBI/ECALL.

Reply:
> If it has an initial value how to distinguish between traps coming from user and those coming from kernel ? Here we are not reading xscratch, we are swapping sp and xscratch, we don't restore xscratch if trap comes from user. Otherwise xscratch is used initially by the boot code to store the hart ID.

@maroueneboubakri I wonder if we can use sstatus.spp to determine if the exception is coming from user mode or Supervisor  mode

Reply:
> > If it has an initial value how to distinguish between traps coming from user and those coming from kernel ? Here we are not reading xscratch, we are swapping sp and xscratch, we don't restore xscratch if trap comes from user. Otherwise xscratch is used initially by the boot code to store the hart ID.
> 
> @maroueneboubakri I wonder if we can use sstatus.spp to determine if the exception is coming from user mode or Supervisor mode

Yes, but I guess it needs temporary space to save some registers and use these registers to determine sstatus.SPP
Otherwise the thread context will be crashed.